src/lib/file-uploader/FileUploader.tsx

```diff
@@ -0,0 +1,202 @@
+// Vendors
+import React, { FC, useEffect } from 'react';
+import { useDropzone, FileRejection, DropEvent, DropzoneOptions } from 'react-dropzone';
+import { get as _get } from 'lodash';
+import classNames from 'classnames';
+
+// Components
+import { ButtonUnstyled } from '../buttons/ButtonUnstyled';
+import { ButtonAccent } from '../buttons/ButtonAccent';
+import { ButtonLink } from '../buttons/ButtonLink';
+import { Icon } from '../icon/Icon';
+
+// Helpers
+import { formatBytes, formatMimeTypesForDisplay } from './FileUploader.helpers';
+
+// Assets
+import placeholderImage from '../../assets/file-uploader-placeholder.png';
+
+// Styles
+import './file-uploader.css';
+
+// Types
+import { FileUpload, FileUploaderProps } from './FileUploader.types';
+
+export const FileUploader: FC<FileUploaderProps> = ({
+  name,
+  className,
+  maxSize = 1048576, // 1 MB
+  minSize,
+  // TODO: Consider allowing file extensions in addition to mime-types
+  accept = ['image/jpeg', 'image/png', 'image/gif'],
+  multiple = false,
+  disabled,
+  dropzoneText,
+  dropzoneActiveText,
+  dropzoneAcceptText,
+  dropzoneRejectText,
+  formikProps,
+  onDrop,
+  onDropAccepted,
+  onDropRejected,
+  onClearAll,
+  onRemoveFile,
+  ...props
+}) => {
+  const files = formikProps ? _get(formikProps?.values, name) || [] : props.value || [];
+  const error = formikProps ? _get(formikProps?.errors, name) : props.error || '';
+
+  useEffect(
+    () => () => {
+      // Make sure to revoke the data uris to avoid memory leaks
+      files.forEach((file: any) => URL.revokeObjectURL(file.preview));
+    },
+    [files]
+  );
+
+  const processFile = (acceptedFile: File): FileUpload => {
+    return Object.assign(acceptedFile, { preview: URL.createObjectURL(acceptedFile) });
+  };
+
+  const processMultipleFiles = (acceptedFiles: File[]): FileUpload[] => {
+    const newFiles = [...(multiple ? files : []), ...acceptedFiles.map((file: any) => processFile(file))];
+
+    // Remove duplicate files
+    const uniqueNewFiles = Array.from(new Set(newFiles.map(a => a.path))).map(path =>
```

Comment:
Do you know how many files we're talking about here? This looks like an O(n^2) operation, mapping across a list of paths and finding their corresponding files.

You could be more efficient converting to a javascript object and letting the keys define uniqueness, assuming I'm understanding this correctly.
```
const uniqueNewFiles = newFiles.reduce((acc, curr) => acc[curr] = curr, {})
return Object.values(uniqueNewFiles)
```

Reply:
@dwene, let me give this a shot and see how it works!